### PR TITLE
Users registration with Trailblazer and friends

### PR DESCRIPTION
**Pull request by bestwebua**

### CODE REVIEW

File: .rspec [hunk 1]
```diff
@@ -0,0 +1,2 @@
+--require spec_helper
+--format documentation
```
Comment (polykoi) [inline .rspec]: Missing newline at end of file

File: .ruby-version [hunk 2]
```diff
@@ -1 +1 @@
-ruby-2.5.0
\ No newline at end of file
+ruby-2.5.3
```
Comment (polykoi) [inline .ruby-version]: Missing newline at end of file

File: Gemfile [hunk 3]
```diff
@@ -1,45 +1,52 @@
...
   gem 'listen', '>= 3.0.5', '< 3.2'
-  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
   gem 'spring-watcher-listen', '~> 2.0.0'
 end
 
-
-# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
-gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
```
Comment (polykoi) [inline Gemfile]: you don't ever need it

File: app/controllers/concerns/api_endpoints.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module ApiEndpoints
+  def default_handler
+    lambda do |model|
+      model.created do |result|
+        render jsonapi: result[:model],
```
Comment (polykoi) [inline app/controllers/concerns/api_endpoints.rb]:
```suggestion
        render jsonapi: result[:model], **result[:renderer_options], status: :created
```
```suggestion
        render jsonapi: result[:model], 
               **result[:renderer_options], 
               status: :created
```
```suggestion
        render jsonapi: result[:model], **result[:renderer_options], 
               status: :created
```

File: app/controllers/concerns/api_endpoints.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        **result[:renderer_options],
+        status: :created
+      end
+
+      model.invalid do |result|
+        render jsonapi_errors: result['contract.default'].errors,
```
Comment (polykoi) [inline app/controllers/concerns/api_endpoints.rb:9]: Wrong indentation

File: app/controllers/v1/application_controller.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module V1
+  class ApplicationController < ActionController::API
+    include JWTSessions::RailsAuthorization
+    rescue_from JWTSessions::Errors::Unauthorized, with: :not_authorized
+
+    private
+
+      def not_authorized
+        render json: { error: 'Not authorized' }, status: :unauthorized
```
Comment (polykoi) [inline app/controllers/v1/application_controller.rb]:
```suggestion
        head: :unauthorized
```

Comment (polykoi) [reply app/controllers/v1/application_controller.rb]: Also you could move `rescue_from` and method to separate concern

File: app/models/todo.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+class Todo < ApplicationRecord
```
Comment (polykoi) [inline app/models/todo.rb]: Ambiguous name

File: config/initializers/mime_types.rb [hunk 8]
```diff
@@ -1,4 +1,7 @@
-# Be sure to restart your server when you modify this file.
+api_mime_types = %w(
+  application/vnd.api+json
```
Comment (polykoi) [inline config/initializers/mime_types.rb]: try leaving only this one

File: config/locales/en.yml [hunk 9]
```diff
@@ -1,33 +1,4 @@
...
-#   'true': 'foo'
-#
-# To learn more, please read the Rails Internationalization guide
-# available at http://guides.rubyonrails.org/i18n.html.
-
 en:
-  hello: "Hello world"
+  errors:
+    file_not_found: 'Documentation api-file not build yet or requested api version not exists.'
+    password_mismatch: 'passwords are different'
```
Comment (polykoi) [inline config/locales/en.yml]: passwords doesn't match

File: lib/tasks/generate_api_documentation.rake [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    task :md do
+      RSpec::Core::RakeTask.new(:api_spec) do |t|
+        t.rspec_opts = "-f Dox::Formatter --order defined --tag dox --out #{md_file}"
+      end
+
+      Rake::Task['api_spec'].invoke
+    end
+
+    task :html do
+      system("aglio -i #{md_file} -o #{html_file}")
```
Comment (polykoi) [inline lib/tasks/generate_api_documentation.rake:17]: prefer ``

File: spec/models/todo_spec.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Todo, type: :model do
+  it { is_expected.to belong_to(:user) }
```
Comment (polykoi) [inline spec/models/todo_spec.rb]: you can assert it to have db fields

File: spec/services/api/version_service_spec.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.describe Api::VersionService do
+  let(:request) do
```
Comment (polykoi) [inline spec/services/api/version_service_spec.rb]: use stub

File: spec/services/api/version_service_spec.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    Class.new do
+      define_method(:headers) do; end
+    end.new
+  end
+
+  let(:headers) { { accept: 'application/vnd.api.v1' } }
+  before { allow(request).to receive(:headers).and_return(headers) }
+
+  describe '#new' do
+    let(:instance) { described_class.new('v1') }
```
Comment (polykoi) [inline spec/services/api/version_service_spec.rb]: move up and use subject

**Review by polykoi — CHANGES_REQUESTED** [review]

File: .rspec (hunk 1 above)
Comment (bestwebua) [reply .rspec]: Fixed.

File: .ruby-version (hunk 2 above)
Comment (bestwebua) [reply .ruby-version]: Fixed.

File: Gemfile (hunk 3 above)
Comment (bestwebua) [reply Gemfile]: Fixed.

File: app/controllers/v1/application_controller.rb (hunk 6 above)
Comment (bestwebua) [reply app/controllers/v1/application_controller.rb]: This controller can be empty for this feature:
```ruby
module V1
  class ApplicationController < ActionController::API
  end
end
```

File: app/controllers/concerns/api_endpoints.rb (hunk 5 above)
Comment (bestwebua) [reply app/controllers/concerns/api_endpoints.rb:9]: Why not!? Works like a charm :)

File: config/initializers/mime_types.rb (hunk 8 above)
Comment (bestwebua) [reply config/initializers/mime_types.rb]: Fixed!

File: config/locales/en.yml (hunk 9 above)
Comment (bestwebua) [reply config/locales/en.yml]: Fixed.

File: lib/tasks/generate_api_documentation.rake (hunk 10 above)
Comment (bestwebua) [reply lib/tasks/generate_api_documentation.rake:17]: Low level style

File: spec/services/api/version_service_spec.rb (hunk 12 above)
Comment (bestwebua) [reply spec/services/api/version_service_spec.rb]: Fixed!

File: spec/services/api/version_service_spec.rb (hunk 13 above)
Comment (bestwebua) [reply spec/services/api/version_service_spec.rb]: Fixed!

**Comment by bestwebua** [comment]
All issues have been fixed.

**Review by polykoi — APPROVED** [review]

**Pull request merged**